openssl/src/cms.rs

```diff
@@ -219,14 +219,66 @@ impl CmsContentInfo {
             Ok(CmsContentInfo::from_ptr(cms))
         }
     }
+
+    /// Verify this CmsContentInfo's signature, given a stack of certificates
+    /// in certs, an X509 store in store. If the signature is detached, the
+    /// data can be passed in data. The data sans signature will be copied
+    /// into output_data if it is present.
+    ///
+    /// OpenSSL documentation at [`CMS_verify`]
+    ///
+    /// [`CMS_verify`]: https://www.openssl.org/docs/manmaster/man3/CMS_verify.html
+    #[allow(clippy::redundant_clone)]
+    pub fn verify(
+        &mut self,
+        certs: Option<&StackRef<X509>>,
+        store: Option<&X509StoreRef>,
+        indata: Option<&[u8]>,
+        output_data: Option<&mut Vec<u8>>,
+        flags: CMSOptions,
+    ) -> Result<(), ErrorStack> {
+        unsafe {
+            let certs_ptr = certs.map_or(ptr::null_mut(), |p| p.as_ptr());
+            if store.is_none() && !flags.contains(CMSOptions::NO_SIGNER_CERT_VERIFY) {
+                panic!("CMS_verify: either the store must be defined or the NO_SIGNER_CERT_VERIFY flag must be set.");
+            }
+
+            let store_ptr = store.map_or(ptr::null_mut(), |p| p.as_ptr());
+            let indata_bio = match indata {
+                Some(data) => Some(MemBioSlice::new(data)?),
+                None => None,
+            };
+            let indata_bio_ptr = indata_bio.as_ref().map_or(ptr::null_mut(), |p| p.as_ptr());
+            let out_bio = MemBio::new()?;
+
+            cvt(ffi::CMS_verify(
+                self.as_ptr(),
+                certs_ptr,
+                store_ptr,
+                indata_bio_ptr,
+                out_bio.as_ptr(),
+                flags.bits(),
+            ))?;
+
+            if let Some(out_data) = output_data {
+                *out_data = out_bio.get_buf().to_vec().clone();
```

Comment:
This clone appears to be unnecessary.

Reply:
I think than out_bio.get_buf() returns an address in the memory managed by OpenSSL which is released when out_bio is dropped. to_vec() creates a new vector from the existing memory (no copy or clone). Therefore the vector needs to be cloned.